### PR TITLE
[Feature] Support RLE for COCO

### PR DESCRIPTION
**Pull request by Indigo6**
## Motivation
Support RLE for COCO and add pre-trained models & log to the model zoo

Arch | Input Size | AP | Total Epochs | Gpu Num
-- | -- | -- | -- | --  
deeppose_resnet_50_rle | 256x192 | 69.35 | 210 | 2

Log and pre-trained model: [Google Drive](https://drive.google.com/drive/folders/12tZjpmkEk3FBR-hvOLG1T2uqWJGnVmBj?usp=sharing)

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by ly015** [comment]
@Indigo6 Thank you very much! We will add other experiment settings (res101, res152) in this PR.

**Comment by Indigo6** [comment]
> @Indigo6 Thank you very much! We will add other experiment settings (res101, res152) in this PR.

You're welcome. I may also open other PRs for coord regression with HRNet, Uncertainty L1/L2 loss

File: mmpose/models/heads/deeppose_regression_head.py [hunk 1]
```diff
@@ -172,6 +172,9 @@ def decode(self, img_metas, output, **kwargs):
 
         preds, maxvals = keypoints_from_regression(output, c, s,
                                                    kwargs['img_size'])
+        if self.out_sigma:
+            maxvals = (1 - output[:, :, 2:].sigmoid())\
```
Comment (Ben-Louis) [inline mmpose/models/heads/deeppose_regression_head.py:176]: Hi, does `output` has the attribute `sigmoid`? It is an np.ndarray according to the docstring.

Comment (Ben-Louis) [reply mmpose/models/heads/deeppose_regression_head.py]: Thanks, it looks good to me now.

Comment (jin-s13) [reply mmpose/models/heads/deeppose_regression_head.py]: I think we need a unit-test for this case. Especially for this bug.

Comment (Ben-Louis) [reply mmpose/models/heads/deeppose_regression_head.py]: Agree. The unittest of `decode` interface is added

**Comment by Indigo6** [comment]
I finished experiments on res101_256x192 but got a result worse than res50.
I look forward to your experimental results and discussion, since there are no official results on res101_256x192.

**Comment by Ben-Louis** [comment]
> I finished experiments on res101_256x192 but got a result worse than res50. I look forward to your experimental results and discussion, since there are no official results on res101_256x192.

Here are the test results of models trained with 8 GPUs

| Arch        | Input Size | AP   |  AP50  | AP75|
| --------   | :-----:  | :-----:  | :----:  |:----:  |
| deeppose_resnet_50_rle| 256x192 | 69.3   |   87.9     | 76.1|
| deeppose_resnet_101_rle| 256x192 |   71.2   |   88.8   | 78.2 |
| deeppose_resnet_152_rle| 256x192 |    72.0    |  88.9  | 79.2 | 

In our experiments, resnet101 performs better than resnet50. 
output

But the AP of resnet50 and resnet152 are worse than that in the paper. We are searching for a better setting for RLE loss on COCO.

**Comment by Indigo6** [comment]
> But the AP of resnet50 and resnet152 are worse than that in the paper. We are searching for a better setting for RLE loss on COCO.

Thanks a lot for sharing these! May I ask for your settings and logs?

**Comment by Ben-Louis** [comment]
> > But the AP of resnet50 and resnet152 are worse than that in the paper. We are searching for a better setting for RLE loss on COCO.
> 
> Thanks a lot for sharing these! May I ask for your settings and logs?

For the three models mentioned above, I use the config you provided. The config files and logs are uploaded to [Google Drive](https://drive.google.com/drive/folders/1GWEq2Zml8Wr29N2gcqV6ecZ7lxpIHka4?usp=sharing). Further results will also be saved in this folder.

**Comment by Indigo6** [comment]
> > > But the AP of resnet50 and resnet152 are worse than that in the paper. We are searching for a better setting for RLE loss on COCO.
> > 
> > 
> > Thanks a lot for sharing these! May I ask for your settings and logs?
> 
> For the three models mentioned above, we use the config you provided. The config files and logs are uploaded to [Google Drive](https://drive.google.com/drive/folders/1GWEq2Zml8Wr29N2gcqV6ecZ7lxpIHka4?usp=sharing). Further results will also be saved in this folder.

Thanks a lot! I'll try again on my devices.
BTW, the reason the APs are worse than those in the paper may be with/without backbone weights pretrained with heatmap loss.

**Comment by Ben-Louis** [comment]
> Thanks a lot! I'll try again on my devices. BTW, the reason the APs are worse than those in the paper may be with/without backbone weights pretrained with heatmap loss.

In the paper, even without pertaining with heatmap loss, the APs of r50/r152 are 70.5/74.2, respectively. They are higher than our results (69.3/72.0). 

The official implementation uses [different hyper-parameters](https://github.com/Jeff-sjtu/res-loglikelihood-regression/blob/master/configs/256x192_res50_regress-flow.yaml) from default setting of deeppose in MMPose, which might be the cause of different performance.

**Comment by Ben-Louis** [comment]
In our experiments, model gains neglectable performance improvement after epoch 210.

| Arch        | Input Size | lr | samples per gpu | data aug | total epoch | AP   |  AP50  | AP75|
| --------   | :-----:  | :-----:  | :----:  |:----:  |:-----:  | :----:  |:----:  |:----:  |
| deeppose_resnet_50_rle| 256x192 | 5r-4 | 64 | MMPose | 210 | 69.3   |   87.9     | 76.1|
| deeppose_resnet_50_rle| 256x192 | 1e-3 | 64 | MMPose | 210 | 69.5   |   87.9     | 76.7|
| deeppose_resnet_50_rle| 256x192 | 1e-3 | 64 | MMPose | 270 | 69.6   |   88.0     | 76.7|
| deeppose_resnet_50_rle| 256x192 | 1e-3 | 32 | Official | 270 | 69.1   |   87.9     | 76.1|

We also tried the setting used in official implementation (batch size, data aug). We found that the loss is lower in training, due to simpler data aug. However the performance reduces with this setting.

**Comment by Indigo6** [comment]
> I finished experiments on res101_256x192 but got a result worse than res50. I look forward to your experimental results and discussion, since there are no official results on res101_256x192.

Two experimental results on my devices

lr | samples per gpu | gpu num | warmup | total epoch | AP | log
:--: | :--: | :--: | :--: | :--: | :--: | :--:
1e-3 | 64 | 2 | no | 210 | 66.91 | [Google Drive](https://drive.google.com/drive/folders/1OG80-hkjffnuxLg8170E_TSP_xIPCNOa)
5e-4 | 64 | 2 | yes | 210 | 71.15 | [Google Drive](https://drive.google.com/drive/folders/1SiRqukmdYScRHDMOS--4DuQaumCIK9Bp)

File: configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192_rle.py [hunk 2]
```diff
@@ -2,11 +2,11 @@
     '../../../../_base_/default_runtime.py',
     '../../../../_base_/datasets/coco.py'
 ]
-evaluation = dict(interval=10, metric='mAP', save_best='AP')
+evaluation = dict(interval=10, metric='mAP', save_best='AP', rle_score=True)
```
Comment (jin-s13) [inline configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192_rle.py:5]: I think it is not a good idea to set `rle_score` in `evaluation` dict. If possible, let's move it to test_cfg.

Comment (jin-s13) [reply configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192_rle.py:5]: I see. That's ok for now. The problem is that the rescoring part is currently in `TopDownCocoDataset.evaluate`. Need refactoring in MMPose 2.0, I think.

Comment (ly015) [reply configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192_rle.py:5]: Yes, we can leave this problem to 2.0 refactoring. Currently, we can add it to the dataset.

**Review by jin-s13 — APPROVED** [review]

**Review by ly015 — APPROVED** [review]
